### DISCUSSION TITLE
How do I get divided screen data?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#7994] [Q&A] [answered]
hello,

I am making RDP on the web using FreeRDP and WebSocket.

When drawing the screen, 

I received the full screen using the structure `context->gdi->primary->bitmap->data` in `mac_end_paint`

However, drawing full screen data uses a lot of resources, so I want to get only the parts that change.

So I used the function `freeerdp_image_copy` in `libfreerdp/codec/color.c`

I succeeded in getting only the changed screen.

However, when you receive the data, you receive the broken screen data together.

The same is true of `gdi_OutputUpdate` in `libfreerdp/gdi/gfx.c`

Part of the screen that `gdi_OutputUpdate` sends. (Color is a mistake)
![image](https://user-images.githubusercontent.com/84696467/175496209-0de49b60-4ad5-46ed-a91e-066a22ea3888.png)

Some of the image data in question

![image](https://user-images.githubusercontent.com/84696467/175496533-646f057b-3712-430a-aa31-d436be988ed9.png)

This is a screen drawn with a split image.
![image](https://user-images.githubusercontent.com/84696467/175496769-ef88454d-2a8e-4a6d-80b5-adf7e6907df7.png)

Is the ability to receive segmented images defined?

Or is there a separate way to draw a segmented image?

I draw a screen using width, height, imagedata, x,y in the function mentioned above.

Thank you for reading the question.

**Comment by akallabeth** [comment, marked as answer]
you most likely ignore the image scanline/stride width? (that is most of the time not `width * bytes_per_pixel` as most of the advanced image routines require some additional alignment)
use that corretly as argument with `freerdp_image_copy` and the glitches should disappear

**Reply by kimcoldwater** [reply]
OK... 

I understand that I have to use scanline/stride.

I tried many things to use it, but I couldn't solve it.

In `freerdp_image_copy`,

I draw images using pDstData.

I made it possible to receive the value from java when `freeerdp_image_copy` is called.


 ```
byte[] data = pDstData; 
int width = dWidth;
int height = dHeight;

BufferedImage image = new BufferedImage(width, height,2);
            Graphics graphics = image.getGraphics();
            int i=0;
            for(int y = 0; y < height; y++){
                for(int x = 0; x < width; x++){

                    Color color = new Color((data[i] & 0xFF),(data[i+1] & 0xFF),(data[i+2] & 0xFF),(data[i+3] & 0xFF));
                    image.setRGB(x,y,color.getRGB());
                    i = i+4;
                }
            }
            graphics.dispose();
```

In Java, uses pDstData ( data ) to create images and send image to a web socket.

Use the for statement to draw the RGBA value of pDstData to width * height.

And I know that the length of this pDstData is stride(nDstStep) * height.

How do I apply scanline/stride to draw RGBA values in this way?



I think the code below in `freerdp_image_copy` is the way to use it, but I don't know the principle.


```
	for (y = 0; y < (INT32)nHeight; y++)
			{
				const BYTE* srcLine =
				    &pSrcData[(y + nYSrc) * nSrcStep * srcVMultiplier + srcVOffset];
				BYTE* dstLine = &pDstData[(y + nYDst) * nDstStep * dstVMultiplier + dstVOffset];
				memcpy(&dstLine[xDstOffset], &srcLine[xSrcOffset], copyDstWidth);
			}

```

**Reply by akallabeth** [reply]
well, how about you check the android client sources? There we copy the native bitmap data to a `AndroidBitmap` (see `Java_com_freerdp_freerdpcore_services_LibFreeRDP_freerdp_1update_1graphics`)

**Reply by kimcoldwater** [reply]
Is there any part of the Mac or Linux module that connects to the freerdp_image_copy in the core module?

Images of size 64 * 64 are normally handed over and suddenly abnormal images are output.

![data7850](https://user-images.githubusercontent.com/84696467/175920426-9fa434b8-b368-45e3-a2c5-568e2eae1c1c.png)

//
![data5335](https://user-images.githubusercontent.com/84696467/175920466-2854b823-b967-4a1f-9c1e-376b1a353346.png)

The two images seem to be the same image, but they are received in a size of 834 * 64 and drawn strangely.

I've been working on it for days, but I can't..😭

**Reply by akallabeth** [reply]
1. Only place you need to take care is `EndPaint` (everything will always be decoded to the `gdi->primary_buffer`)
2. You can either copy the whole image from there or just the rectangles with `freerdp_image_copy`
3. You have lots of examples for that in the `EndPaint` implementations of the clients in the `client` subfolder

**Reply by kimcoldwater** [reply]
Thanks to your advice, we solved it.

I thought I was getting the split image value from a server.

However, we found that 'EndPaint' is a feature that tells us what has changed in the entire image.

I using `gdi->primary->bitmap->data` , `gdi->width`, `gdi->height`, `invalid->x`,`invalid->y`,`invalid->w`, `invalid->h` . 

In EndPaint .

```
  BufferedImage image = new BufferedImage(Rwidth, Rheight,2);
            Graphics graphics = image.getGraphics();
            for(int y = 0; y < Rheight ; y++){
                int i;
                if(Rx == 0 && Ry == 0 ){
                    i= 0;
                }
                else if(Rx == 0){
                    i = width* 4 * Ry;
                }
                else if(Ry == 0){
                    i = Rx * 4;
                }
                else {
                    i = ((Ry - 1) * (4 * width)) + Rx * 4;
                }
                for(int x = 0; x < Rwidth ; x++){
                    Color color = new Color((data[i+2] & 0xFF),(data[i+1] & 0xFF),(data[i] & 0xFF),(data[i+3] & 0xFF));
                    image.setRGB(x,y,color.getRGB());

                    i = i+4;
                }
                Ry++;
            }
            graphics.dispose();
```

I succeeded in getting split image data from Java. Thank you.
